rfcs/45-stageless.md

```diff
@@ -0,0 +1,1014 @@
+        .add_systems(
+            systems![tick_construction_timer, update_construction_progress]
+                .in_set(GameSystems::Construction))
+        .add_system(mitigate_meltdown.run_if(too_many_enemies))
+        // You can use closures for simple one-off conditions.
+        .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+        // `resource_exists` and `resource_equals` are helper functions that produce
+        // new closures that can be used as conditions.
+        .add_system(gravity.run_if(resource_exists(Gravity)))
+        // The systems in this set won't run if `Paused` is `true`.
+        .configure_set(GameSystems::Physics.run_if(resource_equals(Paused(false))))
+        .run();
+}
+```
+
+### Exclusive systems
+
+As much as everyone loves to see systems running in parallel, sometimes a little `&mut World` is necessary.
+Previous versions of Bevy treated these "exclusive" systems as a separate concept, with their own set of types and traits.
+They couldn't have `Local` params and could only be inserted at specific points of a stage.
+
+That is no longer the case.
+Exclusive systems are now just regular systems.
+They can have `Local` params and be scheduled wherever you want.
+
+Command application is now signaled through an exclusive system called `apply_system_buffers`.
+You can add instances of this system anywhere in your schedule.
+If one system depends on the effects of commands from another, make sure an `apply_system_buffers` appears somewhere between them.
+
+```rust
+use bevy::prelude::*;
+
+#[derive(SystemLabel)]
+enum MySystems {
+    /* ... */
+    FlushProjectiles,
+    /* ... */
+}
+
+struct ProjectilePlugin;
+
+impl Plugin for ProjectilePlugin {
+    fn build(app: &mut App) {
+        app
+        .add_system(spawn_projectiles.before(CoreSystems::FlushPostUpdate))
+        .add_systems(
+            chain![
+                check_if_projectiles_hit,
+                despawn_projectiles_that_hit,
+                // Be mindful when using apply_system_buffers in chains;
+                // this will insert an entirely new copy of the system into your schedule.
+                // This can have unexpected effects on system parallelism, as it bifurcates the scheduling graph.
+                apply_system_buffers,
+                fork_projectiles,
+            ]
+            .after(CoreSystems::FlushPostUpdate)
+            .before(CoreSystems::FlushLast)
+        )
+    }
+}
+```
+
+### Running other schedules on-demand with exclusive systems
+
+Exclusive systems can be used to run other schedules in an on-demand fashion.
+Each system is stored within a `Schedule`, which themselves live within a public `Schedules` resource.
+
+```rust
+fn fancy_exclusive_system(world: &mut World) {
+    // removes the selected schedule from Schedules, executes a fn with it, then returns it to the world
+    world.schedule_scope(ScheduleLabel, |world, mut schedule| {

```

Comment:
Actually: are schedule scopes even useful now?

If running the schedule for a given label is the only allowed op, I think we only need `world.run_schedule(ScheduleLabel).`

Reply:
At the very least, I think we should replace all of the instances of `schedule_scope` in this RFC with `run_schedule`.